## 20. Python API - (Application Programming Interface)


### 1. What is an API?

API - A way for one software to talk to another software.


Real-life example:

Phone ---> food delivery app ---> Restaurant

You don't cook, you request food[application] ----> API works the same way.



In Programming:

Client(Your Python Code) requests data 

Servers responds with data.


----

### 2. What is Web API?

A Web API uses the internet(HTTP-Hyper Text Transfer protocol) to exchage data.


Most Common Format:

JSON [JavaScript Object Notation]


Example:


In [4]:
data={
    "name":"Mohan",
    "age":55,
    "course":"Python"
}
data

{'name': 'Mohan', 'age': 55, 'course': 'Python'}

In [3]:
# stringfy

import json

In [6]:
json_string = json.dumps(data,)
json_string

'{"name": "Mohan", "age": 55, "course": "Python"}'

#### 3. API Request Types(HTTP Methods)

-----

|Method|Meaning|Example|
|------|------|------|
|GET|Read data from server|Get user list|
|POST|Send data to server| Create User|
|PUT|Update Major filed Data| Update Data|
|PATCH| Update Minor Filed Data| Update Minor Data|
|DELETE|Remove data from server| Delete Record|

### 5. Required Tools for Python API

install required library



In [7]:
import requests

#### Why requests??

* Simple
* Beginner friedly
* Widely used

### 5. Your First Python API Program (GET Request)

URL- Uniform Resourse Locator

https://jsonplaceholder.typicode.com/users

https://jsonplaceholder.typicode.com/users

code (With Line-by-line)


In [8]:
# step 1: Import Requests Library

import requests

Note:

This library allows Python to send HTTP requests

In [11]:
# step 2: Store API URL


url = "https://jsonplaceholder.typicode.com/users"
url

'https://jsonplaceholder.typicode.com/users'

This is the API endpoints(Server address)

In [32]:
# step3: Send GET request ---> Read data from server

response = requests.get(url)  # GET ---> Reads data from server
print(response)   # if you will get 200 then server response successfully 

<Response [200]>


Note:

* Sends reques to server
* Server replies and stores result in response variable

In [14]:
# step 4 : check status code 

print(response.status_code)

200


Note:

* 200 ---> Success
* 404 ---> Not Found
* 500 ---> Server Error

In [15]:
# step 5: Convert response to JSON

data = response.json()   # string to object 
print(data)

[{'id': 1, 'name': 'Leanne Graham', 'username': 'Bret', 'email': 'Sincere@april.biz', 'address': {'street': 'Kulas Light', 'suite': 'Apt. 556', 'city': 'Gwenborough', 'zipcode': '92998-3874', 'geo': {'lat': '-37.3159', 'lng': '81.1496'}}, 'phone': '1-770-736-8031 x56442', 'website': 'hildegard.org', 'company': {'name': 'Romaguera-Crona', 'catchPhrase': 'Multi-layered client-server neural-net', 'bs': 'harness real-time e-markets'}}, {'id': 2, 'name': 'Ervin Howell', 'username': 'Antonette', 'email': 'Shanna@melissa.tv', 'address': {'street': 'Victor Plains', 'suite': 'Suite 879', 'city': 'Wisokyburgh', 'zipcode': '90566-7771', 'geo': {'lat': '-43.9509', 'lng': '-34.4618'}}, 'phone': '010-692-6593 x09125', 'website': 'anastasia.net', 'company': {'name': 'Deckow-Crist', 'catchPhrase': 'Proactive didactic contingency', 'bs': 'synergize scalable supply-chains'}}, {'id': 3, 'name': 'Clementine Bauch', 'username': 'Samantha', 'email': 'Nathan@yesenia.net', 'address': {'street': 'Douglas Exten

step 1: import API

step 2: url --  endpoint

step 3: GET() ----> Read Data from URL

setp 4: response ---> 200 port number Success

step 5: data = response.json() ---> string --- python Object 

step 6: print(data) -----> Python Object display on screen

### 6. Understanding API Response Structure

In [17]:
print(type(data))

<class 'list'>


each item:


In [20]:
print(type(data[0]))

<class 'dict'>


Notes:

* API response is usually:
  * List of dictionaries
  * Dictionary of dictionaries

### 7. Access Specific Data (Very Important)

In [21]:
for user in data:
    print(user['name'])

Leanne Graham
Ervin Howell
Clementine Bauch
Patricia Lebsack
Chelsey Dietrich
Mrs. Dennis Schulist
Kurtis Weissnat
Nicholas Runolfsdottir V
Glenna Reichert
Clementina DuBuque


In [22]:
for user in data:
    print(user['id'])

1
2
3
4
5
6
7
8
9
10


In [23]:
for user in data:
    print(user['username'])

Bret
Antonette
Samantha
Karianne
Kamren
Leopoldo_Corkery
Elwyn.Skiles
Maxime_Nienow
Delphine
Moriah.Stanton


In [24]:
for user in data:
    print(user['email'])

Sincere@april.biz
Shanna@melissa.tv
Nathan@yesenia.net
Julianne.OConner@kory.org
Lucio_Hettinger@annie.ca
Karley_Dach@jasper.info
Telly.Hoeger@billy.biz
Sherwood@rosamond.me
Chaim_McDermott@dana.io
Rey.Padberg@karina.biz


In [25]:
for user in data:
    print(user['address'])

{'street': 'Kulas Light', 'suite': 'Apt. 556', 'city': 'Gwenborough', 'zipcode': '92998-3874', 'geo': {'lat': '-37.3159', 'lng': '81.1496'}}
{'street': 'Victor Plains', 'suite': 'Suite 879', 'city': 'Wisokyburgh', 'zipcode': '90566-7771', 'geo': {'lat': '-43.9509', 'lng': '-34.4618'}}
{'street': 'Douglas Extension', 'suite': 'Suite 847', 'city': 'McKenziehaven', 'zipcode': '59590-4157', 'geo': {'lat': '-68.6102', 'lng': '-47.0653'}}
{'street': 'Hoeger Mall', 'suite': 'Apt. 692', 'city': 'South Elvis', 'zipcode': '53919-4257', 'geo': {'lat': '29.4572', 'lng': '-164.2990'}}
{'street': 'Skiles Walks', 'suite': 'Suite 351', 'city': 'Roscoeview', 'zipcode': '33263', 'geo': {'lat': '-31.8129', 'lng': '62.5342'}}
{'street': 'Norberto Crossing', 'suite': 'Apt. 950', 'city': 'South Christy', 'zipcode': '23505-1337', 'geo': {'lat': '-71.4197', 'lng': '71.7478'}}
{'street': 'Rex Trail', 'suite': 'Suite 280', 'city': 'Howemouth', 'zipcode': '58804-1099', 'geo': {'lat': '24.8918', 'lng': '21.8984'}

In [26]:
for user in data:
    print(user['phone'])

1-770-736-8031 x56442
010-692-6593 x09125
1-463-123-4447
493-170-9623 x156
(254)954-1289
1-477-935-8478 x6430
210.067.6132
586.493.6943 x140
(775)976-6794 x41206
024-648-3804


In [27]:
for user in data:
    print(user['website'])

hildegard.org
anastasia.net
ramiro.info
kale.biz
demarco.info
ola.org
elvis.io
jacynthe.com
conrad.com
ambrose.net


In [28]:
for user in data:
    print(user['company'])

{'name': 'Romaguera-Crona', 'catchPhrase': 'Multi-layered client-server neural-net', 'bs': 'harness real-time e-markets'}
{'name': 'Deckow-Crist', 'catchPhrase': 'Proactive didactic contingency', 'bs': 'synergize scalable supply-chains'}
{'name': 'Romaguera-Jacobson', 'catchPhrase': 'Face to face bifurcated interface', 'bs': 'e-enable strategic applications'}
{'name': 'Robel-Corkery', 'catchPhrase': 'Multi-tiered zero tolerance productivity', 'bs': 'transition cutting-edge web services'}
{'name': 'Keebler LLC', 'catchPhrase': 'User-centric fault-tolerant solution', 'bs': 'revolutionize end-to-end systems'}
{'name': 'Considine-Lockman', 'catchPhrase': 'Synchronised bottom-line interface', 'bs': 'e-enable innovative applications'}
{'name': 'Johns Group', 'catchPhrase': 'Configurable multimedia task-force', 'bs': 'generate enterprise e-tailers'}
{'name': 'Abernathy Group', 'catchPhrase': 'Implemented secondary concept', 'bs': 'e-enable extensible e-tailers'}
{'name': 'Yost and Sons', 'cat

In [29]:
for user in data:
    print(user['company']['name'])

Romaguera-Crona
Deckow-Crist
Romaguera-Jacobson
Robel-Corkery
Keebler LLC
Considine-Lockman
Johns Group
Abernathy Group
Yost and Sons
Hoeger LLC


In [30]:
for user in data:
    print(user['company']['catchPhrase'])

Multi-layered client-server neural-net
Proactive didactic contingency
Face to face bifurcated interface
Multi-tiered zero tolerance productivity
User-centric fault-tolerant solution
Synchronised bottom-line interface
Configurable multimedia task-force
Implemented secondary concept
Switchable contextually-based project
Centralized empowering task-force


In [31]:
for user in data:
    print(user['company']['bs'])

harness real-time e-markets
synergize scalable supply-chains
e-enable strategic applications
transition cutting-edge web services
revolutionize end-to-end systems
e-enable innovative applications
generate enterprise e-tailers
e-enable extensible e-tailers
aggregate real-time technologies
target end-to-end models


### 8. POST Request (Send Data to API)

Create a new user

In [48]:
# import requests 

url = "https://jsonplaceholder.typicode.com/posts"  # endpoint
url

'https://jsonplaceholder.typicode.com/posts'

In [49]:
payload = {
    "title":"Python API Training",
    "body":"Python is a programming Language",
    "userId":101
}
payload

{'title': 'Python API Training',
 'body': 'Python is a programming Language',
 'userId': 101}

Note:

payload = data we send to server

In [50]:
response = requests.post(url,json=payload)  # convert python object to json string ---> dumps()
print(response)

<Response [201]>


##### The HTTP 201 Created status code is a success response indicating that a new 
##### resource has been successfully created on the server.

Note:

* post() ---> create data in server
* json= payload() ---> auotmatically convertys to JSON()

In [40]:
print(response.status_code)

201


In [42]:
print(response.json())  # convert json_string to python object 

{'title': 'Python API Training', 'body': 'Python is a programming Language', 'userId': 101, 'id': 101}


#### 9. PUT Request (Updata Data) ---> Major data

In [54]:
url = "https://jsonplaceholder.typicode.com/posts/1"

updated_data = {
    "title": "Python Full Course API",
    "body": "Updated Python Content",
    "userId":1
}
# updated_data

response  = requests.put(url, json=updated_data)
print(response)

<Response [200]>


Note:

PUT = replace old data with new data every and each filed

### 10. PATCH --> Request(Update Data) ---> Minor data

In [57]:
url = "https://jsonplaceholder.typicode.com/posts/1"

updated_data = {
    "title": "Python Full Course API",
    "userId":1
}
# updated_data

response  = requests.patch(url, json=updated_data)  # total column = 5, if you wanto to update 3 column then path, if 5 updated then put
print(response)
print(response.status_code)

<Response [200]>
200


### 11. DELETE Request



In [58]:
url = "https://jsonplaceholder.typicode.com/posts/1"

response = requests.delete(url)  # Data deleted from server successfully
print(response.status_code)

200


### 12. Types of APIs

1. Web API (Most Common)
    > Used through internet using HTTP. Example: REST API
2. Library API
    > Example: Python math Module Functions
3. OS API
    > Example: Windows/Linux System Calls
4. Database API
    > Example: MySQL Connector, SQLite API


### API Communication Model
----

API Works on Request -> Response concept

Python Sends:

* URL
* Method
* Data
* Headers


Server Returns:

* Stats Code
* Data (json/xml)
* Headers

### API Protocols

### REST API (Most Used)
* Uses HTTP method
* Uses JSON format


### SOAP API

* Uses XML
* Heavy and secure

### GraphQL API
* Query
* Client chooses requred filed


### REST API Basic

REST --> Representational State Transfer


REST Uses:

* URL - (Endpoint)
* HTTP Method
* JSON Data

Example endpoint:

url- https://api.exmaple.com/user

### HTTP Methods in REST API

---

|Method|Purpose|Example|
|----|----|----|
|GET|Fetch data|Get all users|
|POST|Send/Create data|Add user|
|PUT|Upadte full data| Replace user information|
|PATCH| Update partial Data| Update one filed like email only|
|DELETE| Delete data| Delete user|


### API Status Code


|Code |Meaning|
|----|----|
|200|Success-OK|
|201|Created|
|204|No Content|
|400|Bad Request|
|401|Unauthorized|
|403|Forbidden|
|404|Not Found|
|500|Internal Server Error|


### API Tools Used in Python

Main Library:

✅ requests (Most Popular)

pip install requests

### Sending Header in API

Headers used for:

* Authentication 
* content type
* Token

Example

In [1]:
import requests

In [2]:
url = "https://jsonplaceholder.typicode.com/posts"

headers = {
    "content-Type":"appliction/json"
}

response = requests.get(url,headers = headers)
print(response)

<Response [200]>


In [3]:
url = "https://jsonplaceholder.typicode.com/posts"

headers = {
    "content-Type":"appliction/json"
}

response = requests.get(url,headers = headers)
print(response)
print(response.json())

<Response [200]>
[{'userId': 1, 'id': 1, 'title': 'sunt aut facere repellat provident occaecati excepturi optio reprehenderit', 'body': 'quia et suscipit\nsuscipit recusandae consequuntur expedita et cum\nreprehenderit molestiae ut ut quas totam\nnostrum rerum est autem sunt rem eveniet architecto'}, {'userId': 1, 'id': 2, 'title': 'qui est esse', 'body': 'est rerum tempore vitae\nsequi sint nihil reprehenderit dolor beatae ea dolores neque\nfugiat blanditiis voluptate porro vel nihil molestiae ut reiciendis\nqui aperiam non debitis possimus qui neque nisi nulla'}, {'userId': 1, 'id': 3, 'title': 'ea molestias quasi exercitationem repellat qui ipsa sit aut', 'body': 'et iusto sed quo iure\nvoluptatem occaecati omnis eligendi aut ad\nvoluptatem doloribus vel accusantium quis pariatur\nmolestiae porro eius odio et labore et velit aut'}, {'userId': 1, 'id': 4, 'title': 'eum et est occaecati', 'body': 'ullam et saepe reiciendis voluptatem adipisci\nsit amet autem assumenda provident rerum 

In [7]:
import json
url = "https://jsonplaceholder.typicode.com/posts"

response = requests.get(url)

data = response.json()   # header or json
print(data)

[{'userId': 1, 'id': 1, 'title': 'sunt aut facere repellat provident occaecati excepturi optio reprehenderit', 'body': 'quia et suscipit\nsuscipit recusandae consequuntur expedita et cum\nreprehenderit molestiae ut ut quas totam\nnostrum rerum est autem sunt rem eveniet architecto'}, {'userId': 1, 'id': 2, 'title': 'qui est esse', 'body': 'est rerum tempore vitae\nsequi sint nihil reprehenderit dolor beatae ea dolores neque\nfugiat blanditiis voluptate porro vel nihil molestiae ut reiciendis\nqui aperiam non debitis possimus qui neque nisi nulla'}, {'userId': 1, 'id': 3, 'title': 'ea molestias quasi exercitationem repellat qui ipsa sit aut', 'body': 'et iusto sed quo iure\nvoluptatem occaecati omnis eligendi aut ad\nvoluptatem doloribus vel accusantium quis pariatur\nmolestiae porro eius odio et labore et velit aut'}, {'userId': 1, 'id': 4, 'title': 'eum et est occaecati', 'body': 'ullam et saepe reiciendis voluptatem adipisci\nsit amet autem assumenda provident rerum culpa\nquis hic c

JSON and DATA

----


|Parameter|Meaning|
|----|----|
|json=|Conversts dictionary into json automatically|
|data=| sends raw from data|


In [ ]:
C/c++

Pandas
Numpy
Matplotlib
PowerBI
Tensorflow

API Security TOPICS

-----

1. HTTPS -[Hyper Text Transfer Protocol Secure]

Encrypted communication

2. Authentication

Token/JWT

3. Authorization

Role-based Access

4. Input Validation

Prevent SQL injection

5. Rate Limit

Prevent DDoS -[Distributed Denial-of-Service (DDoS)]

API Documentation

Tools:

* Swagger/Open API
* Postman Documentation
* DRF browsable API -- [Django Rest Framework]


In [ ]:
Tomorrow: Doubt Classes--
----------

Minimal Projecet--
    